llama/model.py

```diff
@@ -203,12 +207,16 @@ def forward(
         output = output.transpose(1, 2).contiguous().view(bsz, seqlen, -1)
         output = self.wo(output)
 
-        # # Activation output sharding
-        # # TODO(yeounoh) remove this after activation sharding support is enabled.
-        num_devices = xr.global_runtime_device_count()
-        device_ids = torch.arange(num_devices)
-        data_model_mesh = xs.Mesh(device_ids, (4, 1, 2))
-        xs.mark_sharding(output, data_model_mesh, (0, 1, 2), use_dynamo_custom_op=True)
+        # Activation output sharding
+        import torch_xla.experimental.dynamo_mark_sharding
+        if self.enable_activation_sharding:
+            # num_devices = 8  # xr.global_runtime_device_count()
+            # device_ids = [i for i in range(num_devices)]
+            device_ids = [0, 1, 2, 3, 4, 5, 6, 7]
```

Comment:
Ah. can we actually pass in the number of devices and have it generates this? or call `num_devices = xr.global_runtime_device_count()` when appropriate

Reply:
We should form the mesh accordingly, [n_devices//2, 1, 2]

Reply:
Updated